graph/pull.go

```diff
@@ -641,20 +640,32 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 		}(&downloads[i])
 	}
 
+	// Register the images serially. We must wait for all downloads to complete if one fails,
+	// in order to cleanup the others.
+	var lastErr error
 	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
-		d := &downloads[i]
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
+		registerFunc := func(d *downloadInfo, pullAborted bool) error {
+			// Cleanup is guranteed to be invoked only after the download routine exits.
+			defer func() {
+				if d.tmpFile != nil {
+					d.tmpFile.Close()
+					os.Remove(d.tmpFile.Name())
+				}
+			}()
+
+			if d.err != nil {
+				if err := <-d.err; err != nil {
```

Comment:
This one was intentional, in order to cleanup the remaining downloads. This will be more critical for #14240.